bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/hooks/ServiceLoaderMediatorHook.java

```diff
@@ -0,0 +1,169 @@
+					return null;
+				}
+			}).filter(Objects::nonNull).limit(15)
+					.filter(c -> c == ServiceLoader.class || c.getEnclosingClass() == ServiceLoader.class).findFirst();
+
+			return () -> {
+				try {
+					return (Optional<Class<?>>) stackWalkerWalk.invoke(stackWalker, function);
+				} catch (ReflectiveOperationException e) {
+					return Optional.empty();
+				}
+			};
+		} catch (ReflectiveOperationException e) { // ignore
+		}
+		try { // Try the Java-1.8 way
+			Class<?> reflectionClass = Class.forName("sun.reflect.Reflection"); //$NON-NLS-1$
+			Method getCallerClass = Objects.requireNonNull(reflectionClass.getMethod("getCallerClass", int.class)); //$NON-NLS-1$
+			return () -> IntStream.range(0, 15).<Class<?>>mapToObj(i -> {
+				try {
+					return (Class<?>) getCallerClass.invoke(null, i);
+				} catch (ReflectiveOperationException e) {
+					return null;
+				}
+			}).filter(Objects::nonNull)
+					.filter(c -> c == ServiceLoader.class || c.getEnclosingClass() == ServiceLoader.class).findFirst();
+
+		} catch (ReflectiveOperationException e) { // ignore an try Java-8 way
+		}
+		throw new AssertionError("Neither the Java-8 nor the Java-9+ way to obtain the caller class is available"); //$NON-NLS-1$
+	}
+
+	private static final String SERVICE_NAME_PREFIX = "META-INF/services/"; //$NON-NLS-1$
+
+	private final ThreadLocal<String> latestServiceName = new ThreadLocal<>();
+
+	private final Supplier<Optional<Class<?>>> callerClass = getCallerClassComputer();
+
+	@Override
+	public Enumeration<URL> preFindResources(String name, ModuleClassLoader classLoader) throws FileNotFoundException {
+		if (name.startsWith(SERVICE_NAME_PREFIX)) {
+			Stream<URL> serviceProviderFiles = searchProviders(name, classLoader, cl -> {
+				try {
+					Enumeration<URL> resources = cl.getResources(name);
+					return Collections.list(resources).stream();
+				} catch (IOException e) {
+					return Stream.empty();
+				}
+			});
+			if (serviceProviderFiles != null) {
+				latestServiceName.set(name);
+				List<URL> services = serviceProviderFiles.collect(Collectors.toList());
+				return Collections.enumeration(services);
+			}
+		}
+		return null;
+	}
+
+	@Override
+	public Class<?> preFindClass(String name, ModuleClassLoader classLoader) throws ClassNotFoundException {
+		String serviceName = latestServiceName.get(); // This cannot be removed if subsequent class load other
+														// providers for the same service
+		if (serviceName != null) {
+			Stream<Class<?>> services = searchProviders(serviceName, classLoader, cl -> {
+				try {
+					return Stream.of(cl.loadClass(name));
+				} catch (ClassNotFoundException | NoClassDefFoundError e) { // ignore
+					return Stream.empty();
+				}
+			});
+			if (services != null) {
+				Optional<Class<?>> findFirst = services.findFirst();
+				return findFirst.orElse(null);
+			}
+		}
+		return null;
+	}
+
+	private <T> Stream<T> searchProviders(String name, ModuleClassLoader classLoader,
```

Comment:
Instead of dynamically searching this list I would use BundleTracker in the Hook that collects bundles and read their `META-INF/services` in a datastructure that can then be used to lookup current active providers, that has several advantages:

1. no need to query the wiring and iterate the bundles
2. easier to debug as there is some current state to inspect
3. instead of the `latestServiceName` we can inspect if the bundle tries to load a class that is an SPI of another bundle and then inspect the stack, requiring a certain call order seems a bit fragile if something changes in the implementation of service loader.

Reply:
> > Instead of dynamically searching this list I would use BundleTracker in the Hook that collects bundles and read their `META-INF/services` in a datastructure that can then be used to lookup current active providers, that has several advantages:
> 
> IIRC the service loader mediator specification requires the impl to look at the wiring to allow a bundle to "use" only the providers it is wired to.

I think so as well. However having a state also allows to avoid the ThreadLocal.
In general I try to have a somewhat hybrid approach that maintains a state with all services, but when there is a query for services checks the current wires if that bundle is permitted to really obtain them.
The goal is also a bit to demonstrate how this could be done totally without (or at least with minimal) metadata as intended in https://github.com/osgi/osgi/issues/372. If that issue is once resolved all the checks for wires could simply be removed and everything possible could be supplied.